### ISSUE TITLE
Update 'openshift service list' command to handle NodePort type

### ISSUE THREAD

**Issue by budhrg** [minishift/minishift#876]
This is a follow up issue of [Update service command to be based on app routes](https://github.com/minishift/minishift/issues/342) to also serve URLs in the format of NodePort.

**Comment by budhrg** [comment]
@hferentschik , need to prioritize it and add to milestone.

**Comment by praveenkumar** [comment]
Currently the way minikube display is quite neat and we can do something similar.

```
$ ./minikube service list
|-------------|----------------------|---------------------------|
|  NAMESPACE  |         NAME         |            URL            |
|-------------|----------------------|---------------------------|
| default     | kubernetes           | No node port              |
| default     | simpleservice        | No node port              |
| kube-system | kube-dns             | No node port              |
| kube-system | kubernetes-dashboard | http://192.168.42.9:30000 |
|-------------|----------------------|---------------------------|

$ ./minishift openshift service list
|-------------|----------------------|---------------------------|
|  PROJECT  |         NAME         |            URL            |
|-------------|----------------------|---------------------------|
| myproject    | kubernetes           | No route              |
| myproject     | simpleservice        | No route              |
|-------------|----------------------|---------------------------|

```

**Comment by hferentschik** [comment]
@praveenkumar, but where in your example is the NodePort. NodePorts and Routes are two different concepts and we want to show both. So we need at least one more column. I think it is wrong to display a NodePort as URL. A NodePort is IP (in our case the IP of the VM) + port. You can expose any sort of protocol via a NodePort. I even argue that if you have HTTP, you most likely will use a Route, whereas if you for example need to expose a database port, you will use a NodePort.

**Comment by praveenkumar** [comment]
@hferentschik I was thinking to only have route and not node port but as @jorgemoralespou suggested in another issue that we can have service and route 2 different sub-commands.

**Comment by hferentschik** [comment]
I think he then agreed to what we have been saying here. I had added a comment and a reference to this issue.

**Comment by praveenkumar** [comment]
> I think he then agreed to what we have been saying here

@hferentschik  Got you. yes than for us output should be look like below right?

```
$ ./minishift openshift service list
|-------------|----------------------|---------------------------|-------------------------|
|  PROJECT  |         NAME           |     NODE-PORT             |  ROUTE                  |
|-------------|----------------------|---------------------------|-------------------------|
```

**Comment by hferentschik** [comment]
something like this I would expect, yes. We would list all services and for each check whether it has a node port and/or route. Depending on that we fill out the appropriate columns.

**Comment by hferentschik** [comment]
There are also other types of services - https://docs.openshift.org/latest/architecture/core_concepts/pods_and_services.html#service-ingressip, but I am not sure whether they are relevant to us. It seems in the context of Minikube and Minishift NodePort and Routes are the interesting things. AFAIU the biggest complaint was also that NodePort disappeared.

**Comment by jorgemoralespou** [comment]
@hferentschik @praveenkumar Remember that a service can have multiple routes and multiple nodePorts, so take that into account.

**Comment by hferentschik** [comment]
Good one, thanks. I guess there are several ways to display this in a table form. We need to discuss the most appropriate one.

**Comment by praveenkumar** [comment]
@jorgemoralespou Currently below are the commands which we are using for display results does that works for multiple routes also or we need to modify this command? 

```
$ oc route -n <namespace> -o=custom-columns=NAME:.metadata.name,HOST:.spec.host 
```

**Comment by jorgemoralespou** [comment]
@praveenkumar depends if you want to make a single table, as displayed above, no, it doesn't work :-D

If you want to show different tables for services and routes, yes, it does work

**Comment by praveenkumar** [comment]
> If you want to show different tables for services and routes, yes, it does work

@jorgemoralespou we can use same sub command and display 2 tables (One for routes and one for services) @hferentschik wdyt?

**Comment by jorgemoralespou** [comment]
I would prefer the 1 table with the valuable information.
Otherwise I would use "oc" commands instead of "minishift". I don't really see the benefit then.

Would a developer trade

````
oc get route
````

for

````
minishift openshift service list
````

and skip the services table?

**Comment by budhrg** [comment]
+1 for one table

**Comment by praveenkumar** [comment]
After looking the way to get node port for a service we need to use go-client for kubernetes because the way it's implemented in minikube is using that client. Or Can we write a custom template to get it?

``` go
   if len(svc.Spec.Ports) > 0 {
		for _, port := range svc.Spec.Ports {
			if port.NodePort > 0 {
				nodePorts = append(nodePorts, port.NodePort)
			}
		}
   }
```

Way I tried is something below but I don't know how now extract nodeport details. @jorgemoralespou any clue?

```
$ kubectl get service -n kube-system -o=custom-columns=PORT:.spec.ports
PORT
[map[name:dns protocol:UDP port:53 targetPort:53] map[name:dns-tcp protocol:TCP port:53 targetPort:53]]
[map[protocol:TCP port:80 targetPort:9090 nodePort:30000]]
```

**Comment by LalatenduMohanty** [comment]
> I would prefer the 1 table with the valuable information.
> Otherwise I would use "oc" commands instead of "minishift". I don't really see the benefit then.

+1. I think this is something we need to keep in mind.

**Comment by LalatenduMohanty** [comment]
> After looking the way to get node port for a service we need to use go-client for kubernetes because the way it's implemented in minikube is using that client. Or Can we write a custom template to get it?

I am confused. Aren't we supposed to use oc command line to extract the information about services, routes and NopePorts? Not sure why we want to use kubectl.

**Comment by hferentschik** [comment]
I am confused as well. For nothing else than _oc_ is a superset of kubectl. I am convinced you can get the information you need from _oc_ as well. If nothing else, try the full YAML or JSON output form _describe_ and parse it.

**Comment by praveenkumar** [comment]
> Not sure why we want to use kubectl.

@LalatenduMohanty it's same for oc also I just put kubectl because was testing those stuff.

```
$ oc get service -n myproject -o=custom-columns=PORT:.spec.ports
PORT
[map[protocol:TCP port:80 targetPort:9876]]
```

**Comment by hferentschik** [comment]
@praveenkumar, so you are saying you want to add a dependency to the Kube Go library. A big -1 from me on this. I cannot believe that we are not able to extract this via _oc_. We finally got rid of this dependency and now you want to re-introduce it?

**Comment by praveenkumar** [comment]
> so you are saying you want to add a dependency to the Kube Go library. A big -1 from me on this. 

@hferentschik I am taking about https://github.com/kubernetes/client-go

**Comment by hferentschik** [comment]
So you are saying the NodePort cannot be extracted from the full YAML description of  a service?

**Comment by praveenkumar** [comment]
> So you are saying the NodePort cannot be extracted from the full YAML description of a service?

Assume a situation where we have more that one service link A,B and C ..., now if we need to first get the service from oc get service => put on the list => iterate over each service => oc describe service_name  => get yaml => parse to get nodePort, In this process we are going to execute lot of time oc commands and handle everything ourself. 

In case of using client library, we just need to create client and query whatever we need so I think thats bit cleaner and client library actively maintained.

**Comment by hferentschik** [comment]
So I was using https://docs.openshift.org/latest/minishift/openshift/exposing-services.html to create an example NodePort. Then I was using _oc decribe/export svc_ and afaict there information is there. In a single YAML really. How hard can it be to iterate this and get the information we need? I don't think we need to run so many _oc_ commands at all. 

I have several issues with this. First our approach and design is to use _oc_. It seems to be wrong to change this for the purpose of just this use case, especially since the information can be retrieved. 

Secondly, client-go looks like a quite bloated library. Adding it will pull all sorts of things back in. This will increase the burden of dependency management, increase build time, ...

**Comment by hferentschik** [comment]
Merged via pull request #1093
